src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -318,22 +319,24 @@ private void HandleCompletionRequest(string id, string paramsJson)
             // Avoid this if possible
             _logger?.Invoke($"[PFX] HandleCompletionRequest: Suggest results: Count:{result.Suggestions.Count()}, Suggestions:{string.Join(", ", result.Suggestions.Select(s => $@"[{s.Kind}]: '{s.DisplayText.Text}'"))}");
 
-            var items = new List<CompletionItem>();
-
-            foreach (var item in result.Suggestions)
+            var precedingCharacter = expression[cursorPosition - 1];
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new
             {
-                items.Add(new CompletionItem()
+                items = result.Suggestions.Select((item, index) => new CompletionItem()
                 {
                     Label = item.DisplayText.Text,
                     Detail = item.FunctionParameterDescription,
                     Documentation = item.Definition,
-                    Kind = GetCompletionItemKind(item.Kind)
-                });
-            }
+                    Kind = GetCompletionItemKind(item.Kind),
 
-            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new
-            {
-                items,
+                    // The order of the results should be preserved.  To do that, we embed the index
+                    // into the sort text, which clients may sort lexigraphically.
+                    SortText = index.ToString(CultureInfo.InvariantCulture),
+
+                    // If the current position is in front of a single quote and the completion result starts with a single quote,
+                    // we don't want to make it harder on the end user by inserting an extra single quote.
```

Comment:
I feel like this logic should be best left up to the client that is integrating Power Fx and not handled here. If I type `'H` and accept the suggestion `'Hello'`, `'Hello'` is still the correct `insertText`, not `Hello'`. This feels more like a client-side issues that should be handled there.

Reply:
This logic is PowerFx specific and should leave in its language server not Monaco. We had a similar problem which looks like got fixed. If you did Connector. and selected the function from intellisense it would insert Connector.Connector.Foobar and not just FooBar. Who is responsible for not inserting Connector here again? Monaco doesn't know what Connector is.

Reply:
To not confuse and block you @NavneetThekkumpat , Merge this pr as you have got the approvals and continue to handle this on the fx bar side. what matters is the end result which is that we need to solve the problem whether that is here or in Fx bar.

Reply:
Discussed in a call - I have come around to Akshar's suggestion and think this does make sense to handle on the Language Server, which has more available context for what the user is doing, particularly in cases when completion is manually triggered.